### DISCUSSION TITLE
5.0.0

### DISCUSSION THREAD

**Discussion by sazzad16** [xetorthio/jedis#3521] [General]
# What's New?

## Automatic Cross-Cluster Failover
We're happy to introduce the Cross-Cluster Failover feature in Jedis. This feature provides high availability and resilience by allowing seamless transitions between Redis clusters during unforeseen failures or downtimes. It's a built-in tool to minimize manual intervention and downtime and ensure a more resilient application infrastructure.
Learn more about how you can automate the failover process in [our documentation](https://github.com/redis/jedis/blob/master/docs/failover.md).

## Full [Redis 7.2](https://redis.com/blog/introducing-redis-7-2/) and [RESP3](https://github.com/redis/redis-specifications/blob/master/protocol/RESP3.md) Support
Examples to enable RESP3 are included later in this release note.

---

# Changes

## 🚀 New Features

- Support RESP3 (#3293, #3387, #3388, #3389, #3403, #3507)
- Support Sharded PubSub (#3396)
- Introduce PipelineBase for Pipeline and multi node pipeline classes (#3437, #3442)
- 'double' timeout parameter for BLMPOP and BZMPOP commands (#3444)

## 🧪 Experimental Features

- Cross cluster failover (#3310)
- Allow setting default dialect for RediSearch module (#3452)
- Support JSON.MERGE command (#3429)
- Support TOPK.LIST with WITHCOUNT option (#3495)

## 🐛 Bug Fixes

- Fix return value of HRANDFIELD with values when count is negative (#3425, #3430)
- Return List instead of Set in ZDIFF, ZINTER, ZUNION commands (#3431)

## 🧰 Maintenance

- Deprecate RedisJSON v1 support (#3503)
- Deprecate RedisGraph support (#3504)
- Deprecate Sharding/Sharded feature (#3386)
- Bump org-json:json from 20230227 to 20230618 (#3472)

---

## RESP3 Examples

This release introduces enabling RESP3 Redis connection, when the Redis server supports it.

1. Enable RESP3 to a UnifiedJedis object:

```java
import redis.clients.jedis.DefaultJedisClientConfig;
import redis.clients.jedis.HostAndPort;
import redis.clients.jedis.UnifiedJedis;

class DoResp3 {
    public static void main() {
        HostAndPort hnp = HostAndPort.from("localhost:6379");
        UnifiedJedis c =  UnifiedJedis(hnp, DefaultJedisClientConfig.builder().resp3().build());
        c.set("foo", "value!");
        c.get("foo");
    }
}
```

2. Enable RESP3 to a Jedis object:

```java
import redis.clients.jedis.DefaultJedisClientConfig;
import redis.clients.jedis.HostAndPort;
import redis.clients.jedis.Jedis;

class DoResp3 {
    public static void main() {
        HostAndPort hnp = HostAndPort.from("localhost:6379");
        Jedis c =  Jedis(hnp, DefaultJedisClientConfig.builder().resp3().build());
        c.set("foo", "value!");
        c.get("foo");
    }
}
```

<hr /><em>This discussion was created from the release <a href='https://github.com/redis/jedis/releases/tag/v5.0.0'>5.0.0</a>.</em>

**Comment by wp973** [comment]
Does jedis(5.0.0) support tracking?

How can jedis client listen the "push" messages? 🧐 @sazzad16

**Reply by sazzad16** [reply]
> Does jedis(5.0.0) support tracking?

No.

> How can jedis client listen the "push" messages?

It doesn't.

**Reply by wp973** [reply]
I'm going to do a POC about **tracking** by **"push"** message, is this possible?

**Reply by wp973** [reply]
Thank you for your reply!

**Reply by sazzad16** [reply]
Possible? Yes.

**Reply by wp973** [reply]
Do you have any plans to support "**tracking**"?

**Reply by wp973** [reply]
I'm thinking of using the approach of two connections mode: https://redis.io/docs/manual/client-side-caching/#two-connections-mode.

Because implementing push messages may break requests/response model ingrained in the source code of the library.

@sazzad16

**Comment by wp973** [comment]
About "**push**" messgae

**Redis conf**: https://redis.io/docs/reference/protocol-spec/#pushes
![image](https://github.com/redis/jedis/assets/65935888/6a95c01e-a9c9-4073-9f47-267b6efed8dc)

**RESP3**: https://github.com/redis/redis-specifications/blob/master/protocol/RESP3.md
![image](https://github.com/redis/jedis/assets/65935888/96a492bb-073b-42ef-80b4-f37fe7660060)

**Comment by chenshi5012** [comment]
like to see aggregation search features in  5.x !!

**Reply by sazzad16** [reply]
@chenshi5012 Please explain. Jedis already has FT.AGGREGATE command support.
It also has `...Iteration` group of methods allowing to aggregate results from all nodes.

**Comment by wbcnr** [comment]
Has the class JedisPoolConfig been deprecated in jedis 5.0.0 version? During code execution: JedisPoolConfig=new JedisPoolConfig (); When, it will report: 

Exception in thread 'main' java. lang. NoSuchMethodError: redis. clients. jedis JedisPoolConfig.setMinEvictableIdleTime(Ljava/time/Duration;) V

Used JDK 1.8

**Reply by sazzad16** [reply]
@wbcnr Your commons-pool version is lower than expected which doesn't have the `setMinEvictableIdleTime(java.time.Duration)` method.
